### PR TITLE
LLL: stricter parser

### PR DESCRIPTION
**Pull request by axic**
Split out off #1298.

### CODE REVIEW

**Comment by axic** [comment]
@zigguratt can you please review the changes?

**Review by zigguratt — APPROVED** [review]
These changes look good.

File: liblll/Parser.cpp [hunk 1]
```diff
@@ -100,16 +100,16 @@ void dev::eth::parseTreeLLL(string const& _s, sp::utree& o_out)
...
 	qi::rule<it, sp::utree()> integer = intstr[qi::_val = px::construct<sp::any_ptr>(px::new_<bigint>(qi::_1))];
 	qi::rule<it, space_type, sp::utree()> atom = integer[qi::_val = qi::_1] | (str | strsh)[qi::_val = qi::_1] | symbol[qi::_val = qi::_1];
 	qi::rule<it, space_type, sp::utree::list_type()> seq = '{' > *element > '}';
-	qi::rule<it, space_type, sp::utree::list_type()> mload = '@' > element;
-	qi::rule<it, space_type, sp::utree::list_type()> sload = qi::lit("@@") > element;
-	qi::rule<it, space_type, sp::utree::list_type()> mstore = '[' > element > ']' > -qi::lit(":") > element;
-	qi::rule<it, space_type, sp::utree::list_type()> sstore = qi::lit("[[") > element > qi::lit("]]") > -qi::lit(":") > element;
-	qi::rule<it, space_type, sp::utree::list_type()> calldataload = qi::lit("$") > element;
-	qi::rule<it, space_type, sp::utree::list_type()> list = '(' > *element > ')';
+	qi::rule<it, space_type, sp::utree::list_type()> mload = '@' > +element;
```
Comment (chriseth) [inline liblll/Parser.cpp]: What does `+element` do?

Comment (zigguratt) [reply liblll/Parser.cpp]: It requires one or more elements to `@`. Before, it accepted a bare `@` in the code, which is an error.

Comment (axic) [reply liblll/Parser.cpp]: That was a remnant from some old changes. `*` means 0 or more, `+` 1 or more, `-` optional and no markup means exactly 1 parameter(s).

**Comment by axic** [comment]
The three changes are:
1. do not accept `0x` as 0, but require at least one digit
2. `()` requires at least one parameter
3. an invalid symbol will not be translated to `PUSH <stacksize>`

For 3. not even @zigguratt can recall the reason.

File: liblll/Parser.cpp (hunk 1 above)
Comment (axic) [reply liblll/Parser.cpp]: Hmm, I've never expected to find proper documentation for this on Wikipedia: https://en.wikipedia.org/wiki/Spirit_Parser_Framework

**Comment by axic** [comment]
Need to verify the problems discovered in https://github.com/ethereum/cpp-ethereum/issues/3387.

**Comment by axic** [comment]
The reason for 3. is that undefined variables become defined at first access and have a value.

`(for [i]:0 (< @i 10) [i](+ @i 1) [j](+ @i @j))`

The undefined variables `i` and `j` will become `0x80` and `0xA0` in the above snippet.

**Comment by chriseth** [comment]
So 3 should be reverted?

**Comment by zigguratt** [comment]
I guess for now it should be reverted. A lot of old LLL code uses auto-defined variables, including the [POC-5 examples](https://github.com/ethereum/cpp-ethereum/wiki/LLL-Examples-for-PoC-5/04fae9e627ac84d771faddcf60098ad09230ab58). But this dubious feature has pretty severe side effects: any typo will auto-assign a variable and no error will be produced. It makes debugging that much more difficult.

But I'll let @axic have the last word regarding reverting 3.

**Comment by zigguratt** [comment]
With regard to @axic's `for` loop example, it would require a simple `(def 'i 0x80) (def 'j 0xa0)` before the `for` loop and then auto-definition is unnecessary and the code becomes clearer.

**Comment by axic** [comment]
We've discussed with @zigguratt that it define-on-use variables potentially cause more harm than benefit, even more so that they exist in the global context.

They can be simply replaced by `def` as written above. Eventually, we should support the `with` keyword from Serpent's LLL to define variables only valid for limited context.

Before merging this change, the cpp-ethereum tests need to be amended to use the def keyword.

**Comment by axic** [comment]
Made the change in cpp-ethereum: https://github.com/ethereum/cpp-ethereum/pull/3401

**Comment by axic** [comment]
@zigguratt can you please double check again this PR against all your LLL source codes?

**Comment by zigguratt** [comment]
@axic I'm out for the day today. I'll have to check tomorrow morning.

**Comment by zigguratt** [comment]
All of my current LLL source compiles on this branch. I think it's ok to go ahead and merge.

**Comment by axic** [comment]
I think we should merge #1442 first.

**Comment by axic** [comment]
Rebased and added tests.

**Comment by axic** [comment]
@chriseth can this be merged? The cpp-ethereum change was merged.

**Review by chriseth — APPROVED** [review]

**Pull request merged**